### PR TITLE
Add snowtable implementation

### PR DESCRIPTION
**Pull request by apcraig**
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Add snowtable implementation
- [X] Developer(s): 
    apcraig
- [ ] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Tested on cheyenne intel with Icepack isnow1tc2 mods.  All results pass but not bit-for-bit with the current consortium master.  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#8a90b84895c3a42806db0345a4183e22715ee066.  Answers here should not be changed relative to current snow1 branch.
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [X] Yes, needs isnow1tc2 mods
    - [ ] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes, but more needs to be done for snow.
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [ ] Please provide any additional information or relevant details below:

The new implementation has the following options

snw_aging_table="test", internally sets the snow table to be the 5x5x1 dataset.

snw_aging_table="snicar", reads the tau, kappa, and drdt0 3D fields from the snw_filename file, reading snw_tau_fname, snw_kappa_fname, and snw_drdt0_fname fields.  The one dimensional fields that support the table lookup are hardwired in icepack (and not totally consistent with the snicar data).

snw_aging_table="file", read the tau, kappa, and drdt0 3D fields from the snw_filename file, reading snw_tau_fname, snw_kappa_fname, and snw_drdt0_fname fields.  In addition, the 1D rhos, Tgrd, and T fields are read from the same file using the snw_rhos_fname, snw_Tgrd_fname, and snw_T_fname field names.

The icepack table lookup only supports tables that have 1D regularly spaces dimensions.  A place holder for non-regularly dimensions is in Icepack, but not yet developed.  Many checks were added to Icepack to ensure the data sizes and regularity are consistent with the implementation.

Minimal documentation on the namelist are added, but more may be needed before merging with master.

### CODE REVIEW

File: cicecore/cicedynB/infrastructure/ice_read_write.F90
Comment (eclare108213) [inline cicecore/cicedynB/infrastructure/ice_read_write.F90:72]: What distinguishes ice_read_nc_2D from _xy?  similarly for 1D/point, 3D/xyz, etc.  Does it make sense to combine any of them?  I also see ice_read_vec_nc, and _uv, and _xyf, above...  These routines are multiplying!  I'm mostly just curious, not asserting that anything has to be done about it at the moment.

File: configuration/scripts/options/set_nml.snw30percent
Comment (eclare108213) [inline configuration/scripts/options/set_nml.snw30percent:2]: snwgrain and use_smliq_pnd should not be set here - my mistake

File: configuration/scripts/options/set_nml.snwgrain
Comment (eclare108213) [inline configuration/scripts/options/set_nml.snwgrain:2]: snwredist should probably not be set here

File: configuration/scripts/options/set_nml.snw30percent
Comment (eclare108213) [reply configuration/scripts/options/set_nml.snw30percent:2]: Both set_nml.nsw30percent and set_nml.nswITDrdg control wind redistribution of snow, which is fundamentally a separate process from the snow grain metamorphism.  What I had in mind is to be able to set the processes individually (with others turned off) as -s snw30percent or -s snwITDrdg, or together as e.g. -s snwITDrdg,snwgrain.  So I don't think we should have snw_aging_table etc in the snow redistribution options, but it should be in snwgrain.  In the case where we want to run both, that should set snw_aging_table appropriately regardless of the order -s snwITDrdg,snwgrain or -s snwgrain,snwITDrdg, right?

File: cicecore/cicedynB/general/ice_forcing.F90
Comment (eclare108213) [inline cicecore/cicedynB/general/ice_forcing.F90:5514]: I'm in favor of just reading everything in, if it's all available in the file anyhow.  What is different in the file from the coded version?  I'm wondering if I don't have the most up-to-date file.

**Review by eclare108213 — APPROVED** [review]
This all looks good to me, though I have some questions below.  For CICE, I think we should only have the 'file' option, not 'snicar'.  For Icepack maybe we should keep both in the code but only offer 'file' officially, so that MPAS can use the 'snicar' approach if necessary.

File: cicecore/cicedynB/infrastructure/ice_read_write.F90
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_read_write.F90:72]: I wanted to use the other implementations, but they are built for CICE arrays specifically and tend to assume sizes and halos.  They are not for general 1D, 2D, 3D arrays, so I had to add those.

File: cicecore/cicedynB/general/ice_forcing.F90
Comment (apcraig) [reply cicecore/cicedynB/general/ice_forcing.F90:5514]: See the icepack PR.  I set the mode to "file" in the set_nml files because I agree it's the right way to do it.  Also, I have not implemented any table reading in Icepack.  That functionality could be ported there, but I thought Icepack was just going to use the "test" data.

File: configuration/scripts/options/set_nml.snwgrain
Comment (apcraig) [reply configuration/scripts/options/set_nml.snwgrain:2]: I leave it to you to review and update the snw set_nml settings.  I updated the defaults, so all the snw flags were basically off.  So the set_nml.snw* files needed to be updated.

Comment (eclare108213) [reply configuration/scripts/options/set_nml.snwgrain:2]: I will change the option file appropriately.

File: cicecore/cicedynB/general/ice_forcing.F90
Comment (eclare108213) [reply cicecore/cicedynB/general/ice_forcing.F90:5514]: Icepack will only use the 'test' data.  If we have both 'snicar' and 'file' in Icepack and remove 'snicar' from CICE, then 'snicar' in Icepack won't get tested.  I'm in favor of removing it completely.  Will wait for N's response.

**Pull request merged**
